File: app/src/main/java/com/example/myintentserviceapp/CustomAdapter.java
package com.example.myintentserviceapp;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityOptionsCompat;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.example.myintentserviceapp.data.Photo;
import com.example.myintentserviceapp.util.RecyclerFastScroller;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.List;

public class CustomAdapter extends RecyclerView.Adapter<CustomAdapter.ViewHolder> implements RecyclerFastScroller.FastScrollable {

    public static final String EXTRA_IMAGE_PATH = "CUSTOM_ADAPTER_IMAGE_PATH";
    public static final String EXTRA_IMAGE_TITLE = "CUSTOM_ADAPTER_IMAGE_TITLE";

    private List<Photo> mList;
    private ViewGroup mParent;
    private int mImageWidth;

    public CustomAdapter(List<Photo> list) {
        this.mList = list;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        mParent = parent;
        mImageWidth = mParent.getWidth() / 4;
        View inflate = LayoutInflater.from(mParent.getContext()).inflate(R.layout.grid_view, parent, false);
        ViewHolder vh = new ViewHolder(inflate);
        return vh;
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        String pathName = mList.get(position).localPath;

        Glide.with(mParent.getContext())
                .load(new File(pathName))
                .override(mImageWidth, mImageWidth)
                .diskCacheStrategy(DiskCacheStrategy.NONE)
                .into(((ViewHolder) holder).myImageView);
    }

    @Override
    public int getItemCount() {
        return mList.size();
    }

    public void add(Photo data) {
        mList.add(data);
    }

    // Method that executes your code for the action received
    public void onItemClick(View view, int position) {
        Photo photo = mList.get(position);

        String filepath = photo.localPath;
        Context context = mParent.getContext();
        Intent intent = new Intent(context, PhotoDetailActivity.class);
        intent.putExtra(EXTRA_IMAGE_PATH, filepath);
        intent.putExtra(EXTRA_IMAGE_TITLE,photo.dateTimeOriginal);

        View imageView = (ImageView) view.findViewById(R.id.card_image_view);
        String sharedElementName = imageView.getTransitionName();
        Activity activity = (Activity) context;
        ActivityOptionsCompat optionsCompat =
                ActivityOptionsCompat.makeSceneTransitionAnimation(activity, imageView, sharedElementName);

        context.startActivity(intent, optionsCompat.toBundle());
        Log.i("TAG", ", which is at cell position " + position + ",You clicked number " + getItem(position).toString());
    }

    private List<Photo> getItem(int position) {
        return mList;
    }

    public void setData(final List<Photo> newPhotos) {
        DiffUtil.DiffResult result = DiffUtil.calculateDiff(new DiffUtil.Callback() {
            @Override
            public int getOldListSize() {
                return mList.size();
            }

            @Override
            public int getNewListSize() {
                return newPhotos.size();
            }

            @Override
            public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                return mList.get(oldItemPosition).sourcePath.equals(newPhotos.get(newItemPosition).sourcePath);
            }

            @Override
            public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                return mList.get(oldItemPosition).createdAt.equals(newPhotos.get(oldItemPosition).createdAt);
            }
        });
        this.mList = newPhotos;
        result.dispatchUpdatesTo(this);
    }

    @NotNull
    @Override
    public String setBubbleText(int position) {
        return mList.get(position).dateTimeOriginal;
    }

    // Stores and recycles views as they are scrolled off screen
    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        public ImageView myImageView;

        public ViewHolder(View itemView) {
            super(itemView);
            myImageView = (ImageView) itemView.findViewById(R.id.card_image_view);

            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            onItemClick(view, getAdapterPosition());
        }
    }
}

File: app/src/main/java/com/example/myintentserviceapp/CustomAdapter.kt
package com.example.myintentserviceapp

import android.app.Activity
import android.content.Intent
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.core.app.ActivityOptionsCompat
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.example.myintentserviceapp.data.Photo
import com.example.myintentserviceapp.util.GlideApp
import com.example.myintentserviceapp.util.RecyclerFastScroller.FastScrollable
import java.io.File

class CustomAdapter(private var mList: MutableList<Photo>) :
    RecyclerView.Adapter<CustomAdapter.ViewHolder>(), FastScrollable {
    private var mParent: ViewGroup? = null
    private var mImageWidth = 0
    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): ViewHolder {
        mParent = parent
        mImageWidth = mParent!!.width / 4
        val inflate =
            LayoutInflater.from(mParent!!.context).inflate(R.layout.grid_view, parent, false)
        return ViewHolder(inflate)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val pathName = mList[position].localPath
        GlideApp.with(mParent!!.context)
            .load(File(pathName))
            .placeholder(R.drawable.ic_baseline_photo_24)
            .error(R.drawable.ic_baseline_error_outline_24)
            .override(mImageWidth, mImageWidth)
            .diskCacheStrategy(DiskCacheStrategy.NONE)
            .into(holder.myImageView)
    }

    override fun getItemCount(): Int {
        return mList.size
    }

    fun add(data: Photo) {
        mList.add(data)
    }

    // Method that executes your code for the action received
    fun onItemClick(view: View, position: Int) {
        val photo = mList[position]
        val filepath = photo.localPath
        val context = mParent!!.context
        val intent = Intent(context, PhotoDetailActivity::class.java)
        intent.putExtra(EXTRA_IMAGE_PATH, filepath)
        intent.putExtra(EXTRA_IMAGE_TITLE, photo.dateTimeOriginal)
        val imageView: View = view.findViewById<View>(R.id.card_image_view) as ImageView
        val sharedElementName = imageView.transitionName
        val activity = context as Activity
        val optionsCompat = ActivityOptionsCompat.makeSceneTransitionAnimation(
            activity,
            imageView,
            sharedElementName
        )
        context.startActivity(intent, optionsCompat.toBundle())
        Log.i(
            "TAG",
            ", which is at cell position " + position + ",You clicked number " + getItem(position).toString()
        )
    }

    private fun getItem(position: Int): List<Photo> {
        return mList
    }

    fun setData(newPhotos: MutableList<Photo>) {
        val result = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
            override fun getOldListSize(): Int {
                return mList.size
            }

            override fun getNewListSize(): Int {
                return newPhotos.size
            }

            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                return mList[oldItemPosition].id == newPhotos[newItemPosition].id
            }

            override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                return mList[oldItemPosition].createdAt == newPhotos[oldItemPosition].createdAt
            }
        })
        mList = newPhotos
        result.dispatchUpdatesTo(this)
    }

    override fun setBubbleText(position: Int): String {
        return mList[position].dateTimeOriginal
    }

    // Stores and recycles views as they are scrolled off screen
    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView),
        View.OnClickListener {
        var myImageView: ImageView
        override fun onClick(view: View) {
            onItemClick(view, adapterPosition)
        }

        init {
            myImageView = itemView.findViewById<View>(R.id.card_image_view) as ImageView
            itemView.setOnClickListener(this)
        }
    }

    companion object {
        const val EXTRA_IMAGE_PATH = "CUSTOM_ADAPTER_IMAGE_PATH"
        const val EXTRA_IMAGE_TITLE = "CUSTOM_ADAPTER_IMAGE_TITLE"
    }
}